module/documents/actor/actor.mjs

```diff
@@ -632,83 +632,111 @@ export default class Actor5e extends Actor {
-        const type = cls.spellcasting.type;
-
-        /**
-         * A hook event that fires while computing the spellcasting progression for each class on each actor.
-         * The actual hook names include the spellcasting type (e.g. `dnd5e.computeLeveledProgression`).
-         * @param {object} progression   Spellcasting progression data. *Will be mutated.*
-         * @param {Actor5e} actor        Actor for whom the data is being prepared.
-         * @param {Item5e} cls           Class for whom this progression is being computed.
-         * @param {number} count         Number of classes with this type of spellcasting.
-         * @returns {boolean}            Explicitly return false to prevent default progression from being calculated.
-         * @function dnd5e.computeSpellcastingProgression
-         * @memberof hookEvents
-         */
-        const allowed = Hooks.call(
-          `dnd5e.compute${type.capitalize()}Progression`, progression, this, cls, types[type] ?? 0
-        );
-
-        if ( allowed && (type === "pact") ) {
-          this.constructor.computePactProgression(progression, this, cls, types.pact);
-        } else if ( allowed && (type === "leveled") ) {
-          this.constructor.computeLeveledProgression(progression, this, cls, types.leveled);
-        }
-      }
+      for ( const cls of classes ) this.constructor.computeClassProgression(
+        progression, this, cls, cls.spellcasting, types[cls.spellcasting.type]
+      );
     }
 
-    const spells = this.system.spells;
     for ( const type of Object.keys(types) ) {
-      /**
-       * A hook event that fires to convert the provided spellcasting progression into spell slots.
-       * The actual hook names include the spellcasting type (e.g. `dnd5e.prepareLeveledSlots`).
-       * @param {object} spells        The `data.spells` object within actor's data. *Will be mutated.*
-       * @param {Actor5e} actor        Actor for whom the data is being prepared.
-       * @param {object} progression   Spellcasting progression data.
-       * @returns {boolean}            Explicitly return false to prevent default preparation from being performed.
-       * @function dnd5e.prepareSpellcastingSlots
-       * @memberof hookEvents
-       */
-      const allowed = Hooks.call(`dnd5e.prepare${type.capitalize()}Slots`, spells, this, progression);
+      this.constructor.prepareSpellcastingSlots(this.system.spells, this, type, progression);
+    }
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Contribute to the actor's spellcasting progression.
+   * @param {object} progression                    Spellcasting progression data. *Will be mutated.*
+   * @param {Actor5e|null} actor                    Actor for whom the data is being prepared.
+   * @param {Item5e} cls                            Class for whom this progression is being computed.
+   * @param {SpellcastingDescription} spellcasting  Spellcasting descriptive object.
```

Comment:
Does this `SpellcastingDescription` need a `typedef`? I can't see one but I might have missed it.

Reply:
Good catch, meant to add that.